File: src/test/java/de/spries/fleetcommander/rest/GameAuthenticatorTest.java
package de.spries.fleetcommander.rest;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

import java.security.GeneralSecurityException;

import org.junit.Before;
import org.junit.Test;

public class GameAuthenticatorTest {

	private static final String INVALID_TOKEN = "invalid token";
	private static final int FIRST_GAME_ID = 1;
	private static final int SECOND_GAME_ID = 2;
	private static final int INEXISTENT_GAME_ID = 12345678;

	private String firstGameToken;

	@Before
	public void setUp() {
		GameAuthenticator.INSTANCE.reset();
		firstGameToken = GameAuthenticator.INSTANCE.createAuthToken(FIRST_GAME_ID);
	}

	@Test(expected = IllegalArgumentException.class)
	public void cannotCreateTokenForGameAgain() throws Exception {
		GameAuthenticator.INSTANCE.createAuthToken(FIRST_GAME_ID);
	}

	@Test
	public void generatedTokenIsValid() throws Exception {
		assertThat(GameAuthenticator.INSTANCE.isAuthTokenValid(FIRST_GAME_ID, firstGameToken), is(true));
	}

	@Test
	public void invalidTokenIsNotValidForGame() throws Exception {
		assertThat(GameAuthenticator.INSTANCE.isAuthTokenValid(FIRST_GAME_ID, INVALID_TOKEN), is(false));
	}

	@Test
	public void generatedTokenIsNotValidForOtherGame() throws Exception {
		String secondGameToken = GameAuthenticator.INSTANCE.createAuthToken(SECOND_GAME_ID);
		assertThat(GameAuthenticator.INSTANCE.isAuthTokenValid(FIRST_GAME_ID, secondGameToken), is(false));
	}

	@Test
	public void tokenIsNotValidForInexistentGame() throws Exception {
		assertThat(GameAuthenticator.INSTANCE.isAuthTokenValid(SECOND_GAME_ID, firstGameToken), is(false));
	}

	@Test
	public void deletingTokenInvalidatesToken() throws Exception {
		GameAuthenticator.INSTANCE.deleteAuthToken(FIRST_GAME_ID, firstGameToken);
		assertThat(GameAuthenticator.INSTANCE.isAuthTokenValid(FIRST_GAME_ID, firstGameToken), is(false));
	}

	@Test(expected = GeneralSecurityException.class)
	public void cannotDeleteTokenForInvalidToken() throws Exception {
		GameAuthenticator.INSTANCE.deleteAuthToken(FIRST_GAME_ID, INVALID_TOKEN);
	}

	@Test(expected = GeneralSecurityException.class)
	public void cannotDeleteTokenForInexistentGameId() throws Exception {
		GameAuthenticator.INSTANCE.deleteAuthToken(INEXISTENT_GAME_ID, firstGameToken);
	}

}

File: src/test/kotlin/de/spries/fleetcommander/web/GameAuthenticatorTest.kt
package de.spries.fleetcommander.web

import de.spries.fleetcommander.web.dto.GamePlayer
import org.hamcrest.Matchers.`is`
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test

class GameAuthenticatorTest {

    private lateinit var firstGamePlayerToken: String
    private val gameAuth = GameAuthenticator()

    @Before
    fun setUp() {
        gameAuth.reset()
        firstGamePlayerToken = gameAuth.createAuthToken(FIRST_GAME_FIRST_PLAYER)
    }

    @Test(expected = IllegalArgumentException::class)
    fun cannotCreateTokenForGameAgain() {
        gameAuth.createAuthToken(FIRST_GAME_FIRST_PLAYER)
    }

    @Test
    fun generatedTokenIsValid() {
        assertThat(gameAuth.isAuthTokenValid(FIRST_GAME_FIRST_PLAYER, firstGamePlayerToken), `is`(true))
    }

    @Test
    fun invalidTokenIsNotValidForGame() {
        assertThat(gameAuth.isAuthTokenValid(FIRST_GAME_FIRST_PLAYER, INVALID_TOKEN), `is`(false))
    }

    @Test
    fun generatedTokenIsNotValidForOtherGame() {
        val secondGameToken = gameAuth.createAuthToken(SECOND_GAME_FIRST_PLAYER)
        assertThat(gameAuth.isAuthTokenValid(FIRST_GAME_FIRST_PLAYER, secondGameToken), `is`(false))
    }

    @Test
    fun generatedTokenIsNotValidForOtherPlayer() {
        val secondGameToken = gameAuth.createAuthToken(FIRST_GAME_SECOND_PLAYER)
        assertThat(gameAuth.isAuthTokenValid(FIRST_GAME_FIRST_PLAYER, secondGameToken), `is`(false))
    }

    @Test
    fun deletingTokenInvalidatesToken() {
        gameAuth.deleteAuthToken(FIRST_GAME_FIRST_PLAYER)
        assertThat(gameAuth.isAuthTokenValid(FIRST_GAME_FIRST_PLAYER, firstGamePlayerToken),
                `is`(false))
    }

    @Test(expected = IllegalArgumentException::class)
    fun cannotDeleteTokenForInexistentGameId() {
        gameAuth.deleteAuthToken(INEXISTENT_GAME_PLAYER)
    }

    companion object {

        private const val INVALID_TOKEN = "invalid token"
        private val FIRST_GAME_FIRST_PLAYER = GamePlayer(1, 1)
        private val FIRST_GAME_SECOND_PLAYER = GamePlayer(1, 2)
        private val SECOND_GAME_FIRST_PLAYER = GamePlayer(2, 1)
        private val INEXISTENT_GAME_PLAYER = GamePlayer(123, 123)
    }

}
